### PR TITLE
depthReduce, a stoppable depth-first reduce for deep iterables

### PR DESCRIPTION
**Pull request by sadasant**
Returns a single item by running a depth-first iteration through a given multi-dimensional iterable (list, object, nested objects and arrays or cyclic object), successively calling the iterator function and passing it an accumulator value, the current value from the iterable, the key where the current value is stored in the iterable, a list with the keys of the parent iterables of the current position, and then passing the result to the next call.

The iterator function receives four values: *(acc, value, key, parentsKeys)*.  It's expected that the iterator returns it's results wrapped in an array, of which only the first position is considered. If the iterator returns an empty array, depthReduce immediately stops, returning the last non-empty result.

See: https://github.com/ramda/ramda/issues/2128#issuecomment-287766254

Please review 🙏🙇

### CODE REVIEW

File: src/depthReduce.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    }
+    var key = env.keys.shift();
+    var value = env.target[key];
+    var result = reducer(accepted, value, key, env.path);
+    if (equals(result, empty)) {
+      return accepted;
+    }
+    accepted = (result && result.length) ? result[0] : result;
+    if (typeof value === 'object') {
+      levels.push({
```
Comment (sadasant) [inline src/depthReduce.js]: 👀 _sublevels_ 🤘 https://open.spotify.com/track/0VwQ3VnCAVT9dqTQa4FYiK 😂

**Comment by sadasant** [comment]
@davidchambers thoughts on this?
[](https://media.giphy.com/media/3oz8xWrNol65TMrc1a/source.gif)

Also, anything else you'd like me to tackle? I see about 175~ issues, I could take some.

File: src/depthReduce.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+ *
+ * Note: `R.depthReduce` does not skip deleted or unassigned indices (sparse
+ * arrays), unlike the native `Array.prototype.reduce` method. For more details
+ * on this behavior, see:
+ * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce#Description
+ *
+ * @func
+ * @memberOf R
+ * @category List
+ * @sig (a,v,k,parentKeys -> [a]) -> a -> [{...}] -> a
```
Comment (sadasant) [inline src/depthReduce.js]: I'm guessing this can be better as `(a,b,c,d -> [a]) -> a -> b -> a`

**Comment by timini** [comment]
??

**Comment by CrossEye** [comment]
@timini: Good `??`  I don't know how this got dropped.  I don't remember ever even seeing it.

@sadasant: I'm very sorry.  I  will try to take a look tomorrow.

**Comment by CrossEye** [comment]
This is a fascinating function that -- like too many old issues/prs was completely dropped.

@sadasant : Do you want to make a new pitch for this?

**Comment by sadasant** [comment]
@CrossEye oh not at all! Sorry for the trouble. Thank you for keeping track of things!

**Comment by CrossEye** [comment]
`Well, the problem is that I wasn't keeping track of things, and neither was anyone else.  @customcommander has joined the team and is helping us get back on track.

I'll close this one now.  If you want to revive it, either comment here or raise a new PR.  Thank you.

And again, sorry.

**Comment by sadasant** [comment]
@CrossEye that sounds like good news! Congrats on onboarding @customcommander ! Ramda is awesome ❤️

**Pull request closed without merging**